File: dashboard/lib/sheets.ts
import { google } from "googleapis";

export type QuoteRow = {
  row: number;
  quote: string;
  author: string;
  status: string;
  source: string;
  created_at: string;
};

async function sheetsClient() {
  const creds = JSON.parse(process.env.GOOGLE_SERVICE_ACCOUNT_JSON!);
  const auth = new google.auth.GoogleAuth({
    credentials: creds,
    scopes: ["https://www.googleapis.com/auth/spreadsheets"],
  });
  return google.sheets({ version: "v4", auth });
}

const RANGE = "Sheet1!A2:E";

export async function listQuotes(status?: string): Promise<QuoteRow[]> {
  const sheets = await sheetsClient();
  const res = await sheets.spreadsheets.values.get({
    spreadsheetId: process.env.GOOGLE_SHEET_ID!,
    range: RANGE,
  });
  const rows = res.data.values ?? [];
  return rows
    .map((r, i) => ({
      row: i + 2,
      quote: r[0] ?? "",
      author: r[1] ?? "",
      status: r[2] ?? "",
      source: r[3] ?? "",
      created_at: r[4] ?? "",
    }))
    .filter((r) => r.quote && (!status || r.status === status));
}

export async function setQuoteStatus(row: number, status: string): Promise<void> {
  const sheets = await sheetsClient();
  await sheets.spreadsheets.values.update({
    spreadsheetId: process.env.GOOGLE_SHEET_ID!,
    range: `Sheet1!C${row}`,
    valueInputOption: "RAW",
    requestBody: { values: [[status]] },
  });
}

File: python/lib/sheets.py
"""Google Sheet as the quote queue. Columns: quote | author | status | source | created_at

status: pending (needs review) -> approved (ready to be carded) -> carded (done)
        or rejected
"""
import json
import os
from datetime import datetime, timezone

import gspread
from google.oauth2.service_account import Credentials

SCOPES = ["https://www.googleapis.com/auth/spreadsheets"]
HEADER = ["quote", "author", "status", "source", "created_at"]


def _worksheet():
    creds_json = os.environ["GOOGLE_SERVICE_ACCOUNT_JSON"]
    creds = Credentials.from_service_account_info(json.loads(creds_json), scopes=SCOPES)
    sheet_id = os.environ["GOOGLE_SHEET_ID"]
    ws = gspread.authorize(creds).open_by_key(sheet_id).sheet1
    if ws.row_values(1) != HEADER:
        ws.update("A1", [HEADER])
    return ws


def append_quotes(quotes: list[dict]) -> None:
    """quotes: [{"quote": str, "author": str, "source": str}]"""
    ws = _worksheet()
    now = datetime.now(timezone.utc).isoformat()
    rows = [[q["quote"], q.get("author", ""), "pending", q.get("source", ""), now] for q in quotes]
    if rows:
        ws.append_rows(rows, value_input_option="RAW")


def list_quotes(status: str | None = None) -> list[dict]:
    ws = _worksheet()
    records = ws.get_all_records()
    out = []
    for i, r in enumerate(records, start=2):  # row 1 is header
        if status is None or r.get("status") == status:
            out.append({**r, "row": i})
    return out


def set_status(row: int, status: str) -> None:
    _worksheet().update_cell(row, HEADER.index("status") + 1, status)
